File: 19/pcm_decode.py
import numpy as np

def pcm_decode(inner, v):


    n = len(inner[0])
    step = np.zeros(8)
    slot = np.zeros(8)
    inner = inner.reshape(8,n//8,order='f')
    slot[0] = 0
    slot[1] = 32
    slot[2] = 64
    slot[3] = 128
    slot[4] = 256
    slot[5] = 512
    slot[6] = 1024
    slot[7] = 2048
    step[0] = 2
    step[1] = 2
    step[2] = 4
    step[3] = 8
    step[4] = 16
    step[5] = 32
    step[6] = 64
    step[7] = 128
    out = np.zeros(n//8)
    for i in range(n//8):
        ss = 2 * inner [0][i] - 1
        tmp = int(inner [1][i] * 4 + inner [2][i] * 2 + inner [3][i] + 1)
        st = slot[tmp]
        dt = (inner [4][i] * 8 + inner [5][i] * 4 + inner [6][i] * 2 + inner [7][i])*step[tmp-1] + 0.5 * step[tmp-1]
        out[i] = ss * (st + dt) / 4096 * v

    return out

File: 19/pcm_decode.m
function [out]= pcm_decode(in,v)
%decode the input pcm code 
%in :   input the pcm code 8bit/sample
%v:     quantized level
n=length(in);
 
in = reshape(in',8,n/8)';
slot(1) = 0;
slot(2) = 32;
slot(3) = 64;
slot(4) = 128;
slot(5) = 256;
slot(6) = 512;
slot(7) = 1024;
slot(8) = 2048;
 
step(1) = 2;
step(2) = 2;
step(3) = 4;
step(4) = 8;
step(5) = 16;
step(6) = 32;
step(7) = 64;
step(8) = 128;
 
 
for i=1:n/8
   ss = 2*in(i,1)-1;
   tmp= in(i,2)*4+in(i,3)*2+in(i,4)+1;
   st = slot(tmp);
   dt = (in(i,5)*8+in(i,6)*4+in(i,7)*2+in(i,8))*step(tmp) + 0.5*step(tmp);
   out(i)=ss*(st+dt)/4096*v;
end
